storage/StorageManager.cpp

```diff
@@ -96,23 +97,60 @@ DEFINE_int32(block_domain, 1,
 static const bool block_domain_dummy
     = gflags::RegisterFlagValidator(&FLAGS_block_domain, &ValidateBlockDomain);
 
-static bool ValidateBufferPoolSlots(const char *flagname,
-                                    std::uint64_t value) {
-  if (value == 0) {
-    std::fprintf(stderr, "--%s must be nonzero\n", flagname);
-    return false;
-  } else {
+/**
+ * @brief Set or validate the buffer pool slots. When automatically picking a
+ *        default value, check if the system is "small" or "large." Set the 
+ *        buffer pool space to 80% of the installed main memory for small
+ *        and 90% otherwise. 
+ *        This method follows the signature that is set by the gflags module.
+ * @param flagname The name of the buffer pool flag. 
+ * @param value The value of this flag from the command line, or default (0)
+ * @return True if the value was set to a legimate value, false otherwise. 
+ *         Currently this method aims to always find some legitimate value,
+ *         and never returns false.
+ **/
+static bool SetOrValidateBufferPoolSlots(const char *flagname,
+                                         std::uint64_t value) {
+  if (value != 0) {
+    // TODO(jmp): Check if this value  is safe and warn the user if it is not.
+    return true;  // User supplied value is > 0 and we simply use that value.
+  }
+
+  // Need to automatically pick the buffer pool size.
+  std::uint64_t total_memory;
+  if (utility::system::calculateTotalMemoryInBytes(&total_memory)) {
+    // Detected the total installed memory. Now set the buffer pool size
+    //   based on whether the system is large or small.
+    if (total_memory/kAGigaByte < kLargeMemorySystemThresholdInGB) {
+      // This is a "small" system. Leave a litte more memory for others.
+      FLAGS_buffer_pool_slots
+        = (total_memory*kPercentageToGrabForSmallSystems)/(kSlotSizeBytes*100);
+    } else {
+      // This is a "large" system. Grab nearly all of the installed memory.
+      FLAGS_buffer_pool_slots
+        = (total_memory*kPercentageToGrabForLargeSystems)/(kSlotSizeBytes*100);
+    }
```

Comment:
Mutiple place states the subsequent line for 4 whitespace indentation:

https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions
https://google.github.io/styleguide/cppguide.html#Constructor_Initializer_Lists